### PR TITLE
allow needless_option_take to report for more cases

### LINKED ISSUE #13671: `Option::as_mut` followed by `Option::take` is a foot-gun
### What it does

Warns users that calling `take` on an `Option` created by `as_mut` does not clear the original `Option`.

It is useless to `take` an `Option` that is only holding a reference, one might as well pattern match on it directly.

### Advantage

It makes it clearer that the original option is not modified.

### Drawbacks

_No response_

### Example

```rust
let mut option = Some("foo");
let maybe_foo = option.as_mut().take();
```

"Calling `take` on an `Option` with a reference does not clear the original `Option`. Remove `as_mut` or pattern-match on the option directly if you did not intend to clear it."

Could be written as:

```rust
let mut option = Some("foo");
let maybe_foo = option.take();
```

```rust
let mut option = Some("foo");
if let Some(a) = option.as_mut() {
    // ...
}
```

### PR DESCRIPTION
**Pull request by ericwu17**
changelog:
```
changelog: [`needless_option_take`]: now lints for all temporary expressions of type  Option<T>.
```

fixes #13671

In general, needless_option_take should report whenever take() is called on a temporary value, not just when the temporary value is created by as_ref().

Also, we stop emitting machine applicable fixes in these cases, since sometimes the intention of the user is to actually clear the original option, in cases such as `option.as_mut().take()`.

### CODE REVIEW

**Comment by ericwu17** [comment]
@Centri3 Are you able to find some time to review this PR or assign another reviewer? Thanks!

**Comment by Centri3** [comment]
I'll be able to look the coming days :+1:

File: clippy_lints/src/methods/needless_option_take.rs [hunk 1]
```diff
@@ -1,30 +1,24 @@
...
-            applicability,
-        );
+    if !recv.is_syntactic_place_expr() && is_expr_option(cx, recv) {
+        if let Some(function_name) = is_creating_temporary_value(recv) {
+            span_lint_and_note(
+                cx,
+                NEEDLESS_OPTION_TAKE,
+                expr.span,
+                "called `Option::take()` on a temporary value",
+                Some(recv.span),
```
Comment (Centri3) [inline clippy_lints/src/methods/needless_option_take.rs:18]: nit: Let's use `None` to allow the note to be inline.

**Review by Centri3 — APPROVED** [review]

File: clippy_lints/src/methods/needless_option_take.rs [hunk 2]
```diff
@@ -1,30 +1,24 @@
...
+                format!("`{function_name}` creates a temporary value"),
```
[the remark this answers is not in the document]
Comment (ericwu17) [reply clippy_lints/src/methods/needless_option_take.rs]: Thanks, I changed the note to the following:

```
`{function_name}` creates a temporary value, so calling take() has no effect.
```

**Comment by ericwu17** [comment]
Hi @Centri3, I have taken a look at your comments and made the appropriate changes. Please let me know if you need anything else from me, and thank you for the review!

**Review by Centri3 — APPROVED** [review]
Thank you.

**Pull request merged**